sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala

```diff
@@ -0,0 +1,127 @@
+import com.nvidia.spark.rapids.jni.ParseURI
+import com.nvidia.spark.rapids.shims.ShimExpression
+
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.rapids.shims.RapidsErrorUtils
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.vectorized.ColumnarBatch
+import org.apache.spark.unsafe.types.UTF8String
+
+object GpuParseUrl {
+  private val HOST = "HOST"
+  private val PATH = "PATH"
+  private val QUERY = "QUERY"
+  private val REF = "REF"
+  private val PROTOCOL = "PROTOCOL"
+  private val FILE = "FILE"
+  private val AUTHORITY = "AUTHORITY"
+  private val USERINFO = "USERINFO"
+}
+
+case class GpuParseUrl(children: Seq[Expression], 
+    failOnErrorOverride: Boolean = SQLConf.get.ansiEnabled) 
+  extends GpuExpression with ShimExpression with ExpectsInputTypes {
+
+  def this(children: Seq[Expression]) = this(children, SQLConf.get.ansiEnabled)
+
+  override def nullable: Boolean = true
+  override def inputTypes: Seq[DataType] = Seq.fill(children.size)(StringType)
+  override def dataType: DataType = StringType
+  override def prettyName: String = "parse_url"
+
+  import GpuParseUrl._
+  
+  override def checkInputDataTypes(): TypeCheckResult = {
+    if (children.size > 3 || children.size < 2) {
+      RapidsErrorUtils.parseUrlWrongNumArgs(children.size) match {
+        case res: Some[TypeCheckResult] => return res.get
+        case _ => // error message has been thrown
+      }
+    }
+    super[ExpectsInputTypes].checkInputDataTypes()
+  }
+
+  def doColumnar(numRows: Int, url: GpuScalar, partToExtract: GpuScalar): ColumnVector = {
+    withResource(GpuColumnVector.from(url, numRows, StringType)) { urlCol =>
+      doColumnar(urlCol, partToExtract)
+    }
+  }
+
+  def doColumnar(url: GpuColumnVector, partToExtract: GpuScalar): ColumnVector = {
+    val part = partToExtract.getValue.asInstanceOf[UTF8String].toString
+    part match {
+      case PROTOCOL =>
+        ParseURI.parseURIProtocol(url.getBase)
+      case HOST | PATH | QUERY | REF | FILE | AUTHORITY | USERINFO =>
+        throw new UnsupportedOperationException(s"$this is not supported partToExtract=$part")
+      case _ =>
+        throw new IllegalArgumentException(s"Invalid partToExtract: $partToExtract")
+    }
+  }
+
+  def doColumnar(url: GpuColumnVector, partToExtract: GpuScalar, key: GpuScalar): ColumnVector = {
+    val part = partToExtract.getValue.asInstanceOf[UTF8String].toString
+    if (part != QUERY) {
+      // return a null columnvector
+      return ColumnVector.fromStrings(null, null)
+    }
+    throw new UnsupportedOperationException(s"$this only supports partToExtract = PROTOCOL")
+  }
+
+  override def columnarEval(batch: ColumnarBatch): GpuColumnVector = {
+    if (children.size == 2) {
+      val Seq(url, partToExtract) = children
+      withResourceIfAllowed(url.columnarEvalAny(batch)) { val0 =>
```

Comment:
If the URL can never be a scalar then don't call columnarEvalAny on it. Just call `columnarEval` which will return a `GpuColumnVector` and use that without the need for the special match.